File: src/bmsx/input/gamepad.ts
import { getPressedState, makeButtonState, resetObject } from './manager';
import type { ButtonState, InputHandler, KeyOrButtonId2ButtonState } from './models';
import type { InputDevice } from '../platform';
import type { VibrationParams } from '../platform';
import { DualSenseHID } from './dualsense_hid';
import { $ } from '../core/engine';


export class GamepadInput implements InputHandler {
	private readonly buttonStates: KeyOrButtonId2ButtonState = {};
	private readonly hidPad = new DualSenseHID();
	private nextPressId = 1;
	private lastPollTime = 0;

	private device: InputDevice;

	constructor(public readonly deviceId: string, public readonly description: string, device: InputDevice) {
		this.device = device;
		this.reset();
	}

	public get gamepadIndex(): number {
		const index = parseInt(this.deviceId.split(':')[1] ?? '-1', 10);
		return index;
	}

	public get supportsVibrationEffect(): boolean {
		return !!this.device?.supportsVibration || this.hidPad.isConnected;
	}

	public setDevice(device: InputDevice): void {
		this.device = device;
	}

	public pollInput(): void {
		const now = $.platform.clock.now();
		const prevPollTime = this.lastPollTime;
		this.lastPollTime = now;

		const keys = Object.keys(this.buttonStates);
		for (let i = 0; i < keys.length; i++) {
			const key = keys[i];
			const state = this.buttonStates[key];
			if (!state) continue;
			if (state.pressed) {
				const pressedAt = state.pressedAtMs ?? state.timestamp ?? now; // TODO: USE resolveStateTimestamp
				state.presstime = Math.max(0, now - pressedAt);
				if (prevPollTime > 0 && state.justpressed && state.timestamp <= prevPollTime) {
					state.justpressed = false;
				}
				state.justreleased = false;
			} else {
				state.presstime = null;
				if (prevPollTime > 0 && state.justreleased && state.timestamp <= prevPollTime) {
					state.justreleased = false;
				}
				state.justpressed = false;
			}
			state.consumed = state.consumed ?? false;
		}
	}

	public getButtonState(btn: string): ButtonState {
		return getPressedState(this.buttonStates, btn);
	}

	public ingestButton(code: string, down: boolean, value: number, timestamp: number, pressId?: number): void {
		const state = this.buttonStates[code] ?? makeButtonState();
		if (down) {
			const existingPressId = pressId ?? state.pressId ?? this.nextPressId++;
			state.pressed = true;
			state.justpressed = true;
			state.justreleased = false;
			state.waspressed = true;
			state.timestamp = timestamp;
			state.pressedAtMs = timestamp;
			state.value = value;
			state.pressId = existingPressId;
		} else {
			const wasPressed = state.pressed;
			state.justreleased = wasPressed;
			state.pressed = false;
			state.justpressed = false;
			state.timestamp = timestamp;
			state.releasedAtMs = timestamp;
			state.value = 0;
			state.waspressed = state.waspressed || wasPressed;
			state.wasreleased = state.wasreleased || wasPressed;
			if (pressId !== undefined) state.pressId = pressId;
			state.consumed = false;
		}
		this.buttonStates[code] = state;
	}

	public ingestAxis2(code: string, x: number, y: number, timestamp: number): void {
		const state = this.buttonStates[code] ?? makeButtonState();
		state.value2d = [x, y];
		state.value = Math.hypot(x, y);
		state.timestamp = timestamp;
		this.buttonStates[code] = state;
	}

	public consumeButton(button: string): void {
		const state = this.buttonStates[button];
		if (state) {
			state.consumed = true;
		}
	}

	public reset(except?: string[]): void {
		if (!except) {
			const keys = Object.keys(this.buttonStates);
			for (let i = 0; i < keys.length; i++) delete this.buttonStates[keys[i]];
			this.lastPollTime = 0;
			return;
		}
		resetObject(this.buttonStates, except);
	}

	public applyVibrationEffect(params: VibrationParams): void {
		if (this.device && this.device.supportsVibration) {
			this.device.setVibration({ effect: 'dual-rumble', duration: params.duration, intensity: params.intensity });
			return;
		}
		if (!this.hidPad.isConnected) {
			const nav = globalThis.navigator as (Navigator & { vibrate?: (pattern: number | number[]) => boolean }) | undefined;
			nav?.vibrate?.(Math.max(0, params.duration * params.intensity));
			return;
		}
		const strongMagnitude = ~~(params.intensity > 0.5 ? params.intensity * 255 : 0);
		const weakMagnitude = ~~(params.intensity <= 0.5 ? params.intensity * 255 : 0);
		try {
			this.hidPad.sendRumble({ strong: strongMagnitude, weak: weakMagnitude, duration: params.duration });
		} catch (err) {
			console.warn(`Error applying HID rumble: ${err}`);
			this.hidPad.disconnect();
		}
	}

	public async init(): Promise<void> {
		try {
			await this.hidPad.initForDevice(this.gamepadIndex, this.description);
		} catch (err) {
			console.warn(`Error initialising HID device for rumble: ${err}`);
		}
	}

	public dispose(): void {
		this.reset();
		this.hidPad.disconnect();
	}
}

File: src/bmsx_cpp/input/gamepad.cpp
/*
 * gamepad.cpp - Gamepad input handler implementation
 */

#include "gamepad.h"
#include "manager.h"
#include "core/console.h"
#include <cmath>

namespace bmsx {

/* ============================================================================
 * Constructor / Destructor
 * ============================================================================ */

GamepadInput::GamepadInput(const std::string& deviceId, const std::string& description)
	: m_deviceId(deviceId)
	, m_description(description)
{
	reset();
}

GamepadInput::~GamepadInput() {
	dispose();
}

/* ============================================================================
 * InputHandler interface
 * ============================================================================ */

void GamepadInput::pollInput() {
	// This is called each frame to update edge flags and press times
	// The actual button state changes come from ingestButton/ingestAxis2
	
	f64 now = ConsoleCore::instance().clock()->now();
	f64 prevPollTime = m_lastPollTimeMs;
	m_lastPollTimeMs = now;
	
	for (auto& [key, state] : m_buttonStates) {
		if (state.pressed) {
			// Update press time
			f64 pressedAt = buttonPressedAtOr(state, now);
			state.presstime = std::max(0.0, now - pressedAt);
			
			// Clear edge flags if they've been processed
			if (prevPollTime > 0 && state.justpressed && 
				buttonTimestampOr(state, 0.0) <= prevPollTime) {
				state.justpressed = false;
			}
			state.justreleased = false;
		} else {
			state.presstime = std::nullopt;
			
			// Clear edge flags if they've been processed
			if (prevPollTime > 0 && state.justreleased &&
				buttonTimestampOr(state, 0.0) <= prevPollTime) {
				state.justreleased = false;
			}
			state.justpressed = false;
		}
	}
}

ButtonState GamepadInput::getButtonState(const ButtonId& button) {
	auto it = m_buttonStates.find(button);
	if (it == m_buttonStates.end()) {
		return ButtonState{};
	}
	return it->second;
}

void GamepadInput::consumeButton(const ButtonId& button) {
	auto it = m_buttonStates.find(button);
	if (it != m_buttonStates.end()) {
		it->second.consumed = true;
	}
}

void GamepadInput::reset(const std::vector<std::string>* except) {
	if (!except) {
		m_buttonStates.clear();
		m_lastPollTimeMs = 0.0;
		return;
	}
	resetObject(m_buttonStates, except);
}

i32 GamepadInput::gamepadIndex() const {
	// Extract index from device ID (format: "gamepad:N")
	size_t colonPos = m_deviceId.find(':');
	if (colonPos == std::string::npos || colonPos + 1 >= m_deviceId.size()) {
		return -1;
	}

	i32 index = 0;
	for (size_t i = colonPos + 1; i < m_deviceId.size(); i++) {
		const char ch = m_deviceId[i];
		if (ch < '0' || ch > '9') {
			return -1;
		}
		index = index * 10 + static_cast<i32>(ch - '0');
	}
	return index;
}

bool GamepadInput::supportsVibrationEffect() const {
	return m_vibrationSupported && m_vibrationCallback;
}

void GamepadInput::applyVibrationEffect(const VibrationParams& params) {
	if (!supportsVibrationEffect()) return;
	m_vibrationCallback(params.intensity, params.duration);
}

void GamepadInput::dispose() {
	reset();
}

/* ============================================================================
 * Button/Axis ingestion
 * ============================================================================ */

void GamepadInput::ingestButton(const std::string& code, bool down, f32 value,
									f64 timestamp, std::optional<i32> pressId) {
	auto& state = m_buttonStates[code];
	
	if (down) {
		i32 existingPressId = resolveButtonPressId(pressId, state, m_nextPressId);
		state.pressed = true;
		state.justpressed = true;
		state.justreleased = false;
		state.waspressed = true;
		state.timestamp = timestamp;
		state.pressedAtMs = timestamp;
		state.value = value;
		state.pressId = existingPressId;
	} else {
		bool wasPressed = state.pressed;
		state.justreleased = wasPressed;
		state.pressed = false;
		state.justpressed = false;
		state.timestamp = timestamp;
		state.releasedAtMs = timestamp;
		state.value = 0.0f;
		state.waspressed = state.waspressed || wasPressed;
		state.wasreleased = state.wasreleased || wasPressed;
		if (pressId.has_value()) {
			state.pressId = pressId;
		}
		state.consumed = false;
	}
}

void GamepadInput::ingestAxis2(const std::string& code, f32 x, f32 y, f64 timestamp) {
	auto& state = m_buttonStates[code];
	state.value2d = Vec2{x, y};
	state.value = std::hypot(x, y);
	state.timestamp = timestamp;
}

} // namespace bmsx
